lib/bridge.js

```diff
@@ -773,52 +783,47 @@ Bridge.prototype._onEvent = function(event) {
         }
     }
 
-    var request = this._requestFactory.newRequest({ data: event });
-    var context = new BridgeContext({
-        sender: event.user_id,
-        target: event.state_key,
-        room: event.room_id
-    });
-    var data = {
+    const request = this._requestFactory.newRequest({ data: event });
+    const contextReady = this._getBridgeContext(event);
+    const dataReady = contextReady.then(c => ({
         request: request,
-        context: context
-    };
-
-    var promise;
-    if (this.opts.disableContext) {
-        promise = Promise.resolve();
-        data.context = null;
-    }
-    else {
-        promise = context.get(this._roomStore, this._userStore);
-    }
+        context: c
+    }));
 
     if (this.opts.queue.perRequest) {
-        promise = Promise.settle([
-            promise,
-            this._prevRequestPromise
-        ]);
+        dataReady = prolong(dataReady, this._prevRequestPromise)
         this._prevRequestPromise = request.getPromise();
     }
 
-    this._queue.push(event, data, promise);
+    this._queue.push(event, dataReady);
     this._queue.consume();
     return request.getPromise();
 };
 
 Bridge.prototype._onConsume = function(err, data) {
-    if (!err) {
-        this.opts.controller.onEvent(data.request, data.context);
+    if (err) {
```

Comment:
Do we not need to check for `this.opts.controller.onLog` ?

Reply:
Afaics it is accounted for in the constructor:

https://github.com/matrix-org/matrix-appservice-bridge/blob/f5fbf9d9c3f36dbad731f6151e234f4ccd6ce705/lib/bridge.js#L156-L163

Reply:
Oh aren't I clever. Yep, that's cool.